ansible_base/lib/routers/association_resource_router.py

```diff
@@ -0,0 +1,156 @@
+        Associate a related object with this object.
+
+        This will be served at /{basename}/{pk}/{related_name}/associate/
+        We will be given a list of primary keys in the request body.
+        """
+        instance = self.parent_view_model.objects.get(pk=kwargs['pk'])
+        serializer = self.get_serializer(data=request.data)
+        serializer.is_valid(raise_exception=True)
+        related_instances = serializer.validated_data['instances']
+        if not related_instances:
+            raise serializers.ValidationError({'instances': _('Please pass in one or more instances to associate')})
+        manager = getattr(instance, self.association_fk)
+        manager.add(*related_instances)
+        return Response(status=status.HTTP_204_NO_CONTENT)
+
+    @action(detail=False, methods=['post'])
+    def disassociate(self, request, **kwargs):
+        """
+        Disassociate a related object from this object.
+
+        This will be served at /{basename}/{pk}/{related_name}/disassociate/
+        We will be given a list of primary keys in the request body.
+        """
+        instance = self.parent_view_model.objects.get(pk=kwargs['pk'])
+        serializer = self.get_serializer(data=request.data)
+        serializer.is_valid(raise_exception=True)
+        related_instances = serializer.validated_data['instances']
+        if not related_instances:
+            raise serializers.ValidationError({'instances': _('Please pass in one or more instances to disassociate')})
+        manager = getattr(instance, self.association_fk)
+
+        # Ensure each of the given related_instances is actually related to the instance.
+        # If any isn't, then bomb out and tell the user which ones aren't related.
+        given_related_instance_set = set(related_instances)
+        related_instance_set = set(manager.all())
+        non_related_instances = given_related_instance_set - related_instance_set
+        if non_related_instances:
+            raise serializers.ValidationError(
+                {
+                    'instances': _('Cannot disassociate these objects because they are not all related to this object: %(non_related_instances)s')
+                    % {'non_related_instances': ', '.join(str(i.pk) for i in non_related_instances)},
+                }
+            )
+
+        try:
+            manager.remove(*related_instances)
+        except AttributeError:
+            # If the field is FK and non-nullable, .remove() won't exist.
+            # https://docs.djangoproject.com/en/dev/ref/models/relations/#django.db.models.fields.related.RelatedManager.remove
+            # "For ForeignKey objects, this method only exists if null=True."
+            # Return something stating that the operation is not supported.
+            raise serializers.ValidationError(
+                {
+                    'instances': _('Cannot disassociate these objects because there must be a related object'),
+                }
+            )
```

Comment:
Is this relevant anymore? Aren't we guaranteed to not be FK now?

Reply:
Yea, I couldn't figure out how to get into this logic from a test. It may just be dead code at this point but I wasn't sure if there was some fringe use case you found during initial development that might bight us if we removed it.